buffer/src/main/java/io/netty/buffer/api/CompositeBuffer.java

```diff
@@ -0,0 +1,2028 @@
+                    "Cannot create composite buffer with duplicate constituent buffer components.");
+            for (Buffer buf : bufs) {
+                try {
+                    buf.close();
+                } catch (Exception closeExc) {
+                    iae.addSuppressed(closeExc);
+                }
+            }
+            throw iae;
+        }
+        return bufs;
+    }
+
+    private static boolean discardEmpty(Buffer buf) {
+        if (buf.capacity() > 0) {
+            return true;
+        } else {
+            // If we filter a buffer out, then we must make sure to close it since it's ownership was sent to us.
+            buf.close();
+            return false;
+        }
+    }
+
+    private static Stream<Buffer> flattenBuffer(Buffer buf) {
+        if (buf instanceof CompositeBuffer) {
+            // Extract components so composite buffers always have non-composite components.
+            var composite = (CompositeBuffer) buf;
+            return Stream.of(composite.bufs);
+        }
+        return Stream.of(buf);
+    }
+
+    private CompositeBuffer(BufferAllocator allocator, Buffer[] bufs, Drop<CompositeBuffer> drop) {
+        super(drop);
+        this.allocator = allocator;
+        try {
+            if (bufs.length > 0) {
+                ByteOrder targetOrder = bufs[0].order();
+                boolean targetReadOnly = bufs[0].readOnly();
+                for (Buffer buf : bufs) {
+                    if (buf.order() != targetOrder) {
+                        throw new IllegalArgumentException("Constituent buffers have inconsistent byte order.");
+                    }
+                    if (buf.readOnly() != targetReadOnly) {
+                        throw new IllegalArgumentException("Constituent buffers have inconsistent read-only state.");
+                    }
+                }
+                order = bufs[0].order();
+                readOnly = targetReadOnly;
+            } else {
+                order = ByteOrder.nativeOrder();
+            }
+            this.bufs = bufs;
+            computeBufferOffsets();
+            tornBufAccessors = new TornBufferAccessors(this);
+        } catch (Exception e) {
+            // Always close bufs on exception, regardless of acquireBufs value.
+            // If acquireBufs is false, it just means the ref count increments happened prior to this constructor call.
+            for (Buffer buf : bufs) {
+                try {
+                    buf.close();
+                } catch (Exception closeExc) {
+                    e.addSuppressed(closeExc);
+                }
+            }
+            throw e;
+        }
+    }
+
+    private void computeBufferOffsets() {
+        if (bufs.length > 0) {
+            int woff = 0;
+            int roff = 0;
+            boolean woffMidpoint = false;
+            for (Buffer buf : bufs) {
+                if (buf.writableBytes() == 0) {
+                    woff += buf.capacity();
```

Comment:
Should this be Math.addExact() ?

Reply:
No, we have a separate check below to ensure we never exceed max possible buffer capacity.